README.md
# NNbenchmark

NNbenchmark was created during the Google Summer of Code, 2019 as a part of The R Project for Statistical Computing, to verify the convergence of the training algorithms provided in 69 Neural Network R packages available on CRAN to date. Neural networks must be trained with second order algorithms and not with first order algorithms as many packages seem to do. 


The purpose of this project is to verify the quality of the training algorithms in R packages that provide neural network of perceptron type (one input layer, one normalized layer, one hidden layer with nonlinear activation function usually tanh(), 
one normalized layer, one output output layer) for regression purpose i.e. NN(X1, ..., Xn) = E[Y].


## Installation

```
install.packages('NNbenchmark')
```

## Documentation

[Reference manual on CRAN](https://cran.r-project.org/web/packages/NNbenchmark/NNbenchmark.pdf).


## Packages Tested  

This GSoC project will conduct a comprehensive survey of all packages that have the “neural network” keyword in thepackage title or in the package description. There are currently 69 packages on CRAN with this keyword.  

|    []()       |               |               |               |               |               |
|:-------------:|:-------------:|:-------------:|:-------------:|:-------------:|:-------------:| 
|   AMORE       |    ANN2       |  appnn        |   automl      |  brnn         |  CaDENCE      |
|   DALEX2      |    DamiaNN    |  deepnet      |   elmNNrcpp   |   ELMR        | EnsembleBase  |                
|   h2o         |   keras       |  MachineShop  |   minpack.lm  |  monmlp       |  neuralnet    |
|  nlsr         |  nnet         |  qrnn         | radiant.model | rminer        | RSNNS         | 
| snnR          |  TraineR      | validann      |



## Evaluation Criteria
***
The algorithms were tested on 12 regression datasets (both univariate and multivariate) of varying complexity.  

The score for each package was based on the following parameters:  

* Documentation (0-3 stars or a binary value of 0/1)
* UtilFunction (0-3 stars or a binary value of 0/1)
* ConvergenceQuality (0-3 stars based on percentile method)
* ConvergenceTime (0-3 stars based on percentile method)


To obtain the final rating, we take a weighted average of these 4 columns (giving more weight to ConvergenceQuality and ConvergenceTime).


## Example Usage

An example usage using the `nnet` package.

### Load packages

```
library(NNbenchmark)
library(kableExtra)
options(scipen = 999)
```

### Datasets to test

```
NNdataSummary(NNdatasets)
```

### Setup nnet `trainPredict` arguments

The trainPredict functions of NNbenchmark were made to be generic so they could accomodate the many different formats (with regards to data preparation, training functions, etc) of neural network packages in R. 
There are 3 versions of trainPredict available.  
* In trainPredict_1mth1data, a neural network is trained on one dataset and then used for predictions, with several functionalities. Then, the performance of the neural network is summarized.
* trainPredict_1data is a wrapper function of trainPredict_1mth1data for multiple methods/algorithms.
* trainPredict_1pkg is a wrapper function of trainPredict_1mth1data for multiple datasets.

The following example shows how to use trainPredict_1pkg to benchmark the nnet package against 12 datasets.

```
## Setup directory
if(dir.exists("D:/GSoC2020/Results/2020run03/"))
{  
  odir <- "D:/GSoC2020/Results/2020run03/"
}else if(dir.exists("~/Documents/recherche-enseignement/code/R/NNbenchmark-project/NNtempresult/"))
{  
  odir <- "~/Documents/recherche-enseignement/code/R/NNbenchmark-project/NNtempresult/"
}else
  odir <- "~"

## Setup number of repetitions, hyperparameters, and algorithm
nrep <- 10
maxit2ndorder  <-    200
maxit1storderA <-   1000
maxit1storderB <-  10000
maxit1storderC <- 100000

nnet.method <- "none"
hyperParams.nnet <- function(...) {
    return (list(iter=maxit2ndorder, trace=FALSE))
}

## Setup data preparation (nnet.prepareZZ), training function, prediction function, and close function
NNtrain.nnet <- function(x, y, dataxy, formula, neur, method, hyperParams, ...) {
    
    hyper_params <- do.call(hyperParams, list(...))
    
    NNreg <- nnet::nnet(x, y, size = neur, linout = TRUE, maxit = hyper_params$iter, trace=hyper_params$trace)
    return(NNreg)
}
NNpredict.nnet <- function(object, x, ...)
    predict(object, newdata=x)
NNclose.nnet <- function()
  if("package:nnet" %in% search())
    detach("package:nnet", unload=TRUE)
nnet.prepareZZ <- list(xdmv = "d", ydmv = "v", zdm = "d", scale = TRUE)
if(FALSE)
res <- trainPredict_1data(1, nnet.method, "NNtrain.nnet", "hyperParams.nnet", "NNpredict.nnet", 
                               NNsummary, "NNclose.nnet", NA, nnet.prepareZZ, nrep=5, echo=TRUE, doplot=FALSE,
                               pkgname="nnet", pkgfun="nnet", csvfile=TRUE, rdafile=TRUE, odir=odir)
```

### Launch `trainPredict`

```
res <- trainPredict_1pkg(1:12, pkgname = "nnet", pkgfun = "nnet", nnet.method,
  prepareZZ.arg = nnet.prepareZZ, nrep = nrep, doplot = TRUE,
  csvfile = TRUE, rdafile = TRUE, odir = odir, echo = FALSE)
```


### Result

```
kable(t(apply(res, c(1,4), min)))%>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed"))
```


## Authors  

### Selected Students:

- Akshaj Verma
- Salsabila Mahdi

### Mentors:

- Patrice Kiener  
- Christophe Dutang
- John C. Nash


NEWS
= Package NNbenchmark =
- Akshaj Verma, Salabila Mahdi, Patrice Kiener
- Christophe Dutang, John Nash 

NNbenchmark 3.2.1 - 2023-09-
==================================================
- start parallelization for trainPredict_1pkg() (inspired from boot pkg)
- add extra-argument lib.loc in trainPredict_1pkg() and trainPredict_1data()


NNbenchmark 3.2.0 - 2021-06-05 - Submitted to CRAN
==================================================
- NNpredict.R:  
- trainPredict_1mth1data(..., odir = "."). Before was "~/" which is incorrect 
- trainPredict_1data(..., odir = ".")      (as path have no final slash in R)
- trainPredict_1pkg(..., odir = ".")
- trainPredict_1mth1data() file.path(odir, ...) replaces paste0(odir, ...)

NNbenchmark 3.1.9 - 2021-03-09 - Submitted to CRAN
==================================================
- All URLs tested with the new package urlchecker_1.0.0 published on 2021-03-04
- datasets.R: http replaced by https
- datasets.R: NIST pages replaced by a redirection from InModelia's website
- DESCRIPTION and a_intro.R: New links to theairbend3r, Github and InModelia
- DESCRIPTION: Akshaj and Salsabila Orcid numbers 

NNbenchmark 3.1.6 - 2020-09-07 - Submitted to CRAN
==================================================
- v_3.1.5 erased since mixed with 2019 code
- v_3.1.6 copied from v_3.1.4

NNbenchmark 3.1.4 - 2020-08-30
==============================
- timediff() updated time0 <- NA ; time0 <<- time0 detected by CRAN
- Authors@R modified: correct PK email
- Author: orcid number for PK and CD

NNbenchmark 3.1.3 - 2020-08-30
==============================
- timediff() updated time0 <<- NA
- NNSummary() and NNtrainPredict() cat(), print() modified

NNbenchmark 3.1.1 - 2020-08-26
==============================
- DESCRIPTION: add Authors@R
- clearNN() and detachNN() further simplified

NNbenchmark 3.1.0 - 2020-08-24 - Submitted to CRAN
==================================================
- change version number and sent to CRAN

NNbenchmark 3.0.7 - 2020-08-24
==============================
- simplify clearNN()
- Christophe Dutang added as author
- Email address: rpackages--@--inmodelia
- false link replaced by the table header "Packages" in README.md
- NNtrainPredict.R: odir <- tempdir() rather than getwd()

NNbenchmark 3.0.5 - 2020-07-30
==============================
- allow for nrep = 1
- added examples to NNtrainPredict
- changed echoreport to logical

NNbenchmark 3.0.4 - 2020-07-03
==============================
- remove commented traces
- add an argument to trainPredict_1mth1data()
- add mWoodG1 rda file (Simon Wood's GAM test file for Gaussian family)

NNbenchmark 3.0.3 - 2020-06-29
==============================
- NNtrainPredict now has the options for csv outputs, same as 2019 legacy code
- round time objects in NNsummary

NNbenchmark 3.0.2 - 2020-06-29
==============================
- adjusted NNtrainPredict and NNsummary functions to new timing functions


NNbenchmark 3.0.1 - 2020-06-27
==============================
- DESCRIPTION: Roxygen: list(old_usage = TRUE)
- timestart(), timediff()


NNbenchmark 3.0.0 - 2020-06-21 (Bila + Patrice)
===============================================
- NNTrain_Predict.R for NEW 2020 code do.call() implementation to train and predict NNs
- Move NNbenchmark2020 to top directory. This is new NNbenchmark
- Move NNbenchmark2019 to a new directory
- Fix the date in DESCRIPTION

NNbenchmark 2.3.7 - 2020-06-19 (Bila)
=====================================
- Added MSE and WAE to funRMSE.R
- Made a new R file: NNSummary.R for 2020 summary function

NNbenchmark 2.3.6 - 2020-01-11
==============================
- ht() and ht9() fully rewritten. Accept now array and data.frame with matrices inside
- README.md added

NNbenchmark 2.3.5 - 2019-08-22
==============================
- DESCRIPTION and a_intro.R: add the html links to GitHub and AkshajVerma.com

NNbenchmark 2.3.4 - 2019-08-21 - sent to CRAN
=============================================
- DESCRIPTION: Modify the title to match title case (recommended by CRAN)

NNbenchmark 2.3.3 - 2019-08-21
==============================
- DESCRIPTION: Modify the title 
- All files: Add value to Rd files (return in roxygen comments) for every functions

NNbenchmark 2.3.2 - 2019-08-20
==============================
- prepareZZ: extend donttest to run examples in less than 5 seconds

NNbenchmark 2.3.1 - 2019-08-20
==============================
- add2csv: examples rewritten with tempfile()
- datalist rewritten with format (file_without_ext: obj1, ..., objn)
- use for this purpose the hidden function tools:::list_data_in_pkg()

NNbenchmark 2.3 - 2019-08-19
============================
- NNdatasets: Change uNeuroOne from 3 to 2 neurons and 10 to 7 parameters
- NNdatasets: mRef153: correct a typo in the number of neurons (from 4 to 3)
- u+mNNdatasets: Update the documentation with the number of parameters for each dataset
- prepareZZ: rearrange the examples with donttest to run less than 5 seconds
- DESCRIPTION: Yifu Yan as author of the timeR function + class (timeR package)
- Test CRAN check and prepare the upload to CRAN

NNbenchmark 2.2 - 2019-08-13
============================
- NNdataSummary()

NNbenchmark 2.1.1 - 2019-08-13
==============================
- clearNN() donotdetach now includes kableExtra, dplyr, stringr

NNbenchmark 2.1 - 2019-08-12
============================
- detachNN() (to move library(pkg) out of the loop in NNTemplate)

NNbenchmark 2.0.1 - 2019-08-04
==============================
- Date 2019-08-04 added at the first line of *.R files
- Linked to GitHub. Package is rebuilt before commit (to force change in NAMESPACE)

NNbenchmark 2.0 - 2019-08-04
============================
- column 'params' added to the timer table. Columns in the data.frame are:
- timer$stop(event, start, end, duration, RMSE, MAE, stars, params, comment)
- Connect to GitHub

NNbenchmark 1.9 - 2019-08-03
============================
- New disk

R/timer_class.R
## timer_class 2019-08-04




#' A R6 Class to represent a timer.
#'
#' timer is a R6 Class that represent a timer. This is a modified version of the 
#' \code{timeR} package for an internal use. Full credit is to Yifu Yan, the author 
#' of the \code{timeR} package.
#' 
#' @return
#' \code{getTimer} returns a data frame with all records saved by the timer 
#' object. Columns in the data.frame are: event, start, end, duration, RMSE, 
#' MAE, stars, params, comment.
#'
#' @docType class
#' @field time A POSIXct/POSIXlt value of your latest timing.
#' @field event A string of your latest timing.
#' @field eventTable A data frame that stores all timings.
#' @field verbose A printing setting that controls whether to print messages.
#' @section Public Methods:
#' \describe{
#'   \item{\code{initialize(time,event,verbose,eventTable)}
#'   }{Initialize a timer object. You can also use \code{createTimer()}
#'   function to initialize a timer object.}
#'   \item{\code{start(eventName)}
#'   }{Start timing for a event, \code{eventName} should be a string}
#'   \item{\code{stop(eventName)}
#'   }{Stop timing for a event.}
#'   \item{\code{getTimer()}
#'   }{Get/Print a data.frame with all records.}
#'   \item{\code{removeEvent(eventName)}
#'   }{Remove an given row in the eventTable.}
#'   \item{\code{toggleVerbose()}
#'   }{Toggle between \code{TRUE} and \code{FALSE} for \code{verbose}}
#'   \item{\code{getStartTime()}
#'   }{Get start time for a selected event.}
#'   \item{\code{getStopTime()}
#'   }{Get stop time for a selected event.}
#'   \item{\code{getDuration()}
#'   }{Get duration for a selected event.}
#'   \item{\code{getRMSE()}
#'   }{Get the RMSE for a selected event.}
#'   \item{\code{getMAE()}
#'   }{Get the MAE for a selected event.}
#'   \item{\code{getStars()}
#'   }{Get stars for a selected event.}
#'   \item{\code{getParams()}
#'   }{Get params for a selected event.}
#'   \item{\code{getComment()}
#'   }{Get comment for a selected event.}
#'   \item{\code{getEventf()}
#'   }{Get entire row for a selected event.}
#'   \item{\code{print()}
#'   }{Custom print method for timer class. However, you don't need to use this
#'   function to generate custom printing.
#'   Custom printing is triggered by default.}
#'   }
#' @section Private Methods:
#' \describe{
#' \item{\code{slprint(msg, flag = self$verbose)}
#' }{A function that controls whether to print extra message.}
#' }
#' @examples
#' timer <- createTimer()
#' timer$start("event1")
#' ## put some codes in between, for instance
#' Sys.sleep(1)
#' timer$stop("event1", RMSE = 1, MAE = 1.3, stars = "*", 
#'            params = "maxiter=100, lr=0.01", comment = "OK for 1",  
#'            printmsg = TRUE)
#'
#' timer$start("event2")
#' ## put some codes in between, for instance
#' Sys.sleep(2)
#' timer$stop("event2", RMSE = 2, MAE = 2.6, stars = "**",  
#'            params = "maxiter=1000, lr=0.001", comment = "OK for 2",  
#'            printmsg = FALSE)
#'
#' table1 <- getTimer(timer)
#' timer$toggleVerbose() # set verbose to FALSE as default is TRUE
#' table1 # print all records in a data frame
#' 
#' ## get attributes for selected events
#' timer$getStartTime("event1")
#' timer$getStopTime("event1")
#' timer$getDuration("event1")
#' timer$getComment("event1")
#' timer$getEvent("event1")
#' 
#' @importFrom R6 R6Class
#' @export
timeR <- R6::R6Class(
    classname = "timeR",
    public = list(
        #values
        eventTable = data.frame(
            event = character(),
            start = character(),
            end = character(),
            duration = numeric(),
            RMSE = numeric(),
            MAE  = numeric(),
            stars = character(),
            params = character(),
            comment = character(),
            stringsAsFactors = FALSE
            ),
        verbose = logical(),
        #initialize timeR
        initialize = function(verbose=TRUE){
            stopifnot(is.logical(verbose),!is.na(verbose))
            self$verbose = verbose
        },
        #start a timeR for event
        start = function(eventName){
            if(!exists("eventName")){
                stop("Please create a name for the event.")
            }
            theTable <- self$eventTable
            verbose <- self$verbose
            current_time <- format(Sys.time(), "%H:%M:%OS")
            # current_time <- as.character(lubridate::now())
            #create that record/row
            newRow <- data.frame(event = eventName,
                                 start = current_time,
                                 end = character(1),
                                 duration = numeric(1),
                                 stringsAsFactors = FALSE,
                                 RMSE = NA_real_,
                                 MAE  = NA_real_,
                                 stars = NA_character_,
                                 params = NA_character_,
                                 comment = NA_character_
                                 )
            #detect ifevent already exist
            if (any(theTable$event %in% eventName) ){
                out_msg <- paste0("Event: '",
                                  eventName,
                                  "' already exists.",
                                  " Overwriting previous 'start'.\n")
                private$slprint(out_msg)
                #replace the row in dataframe with new row
                boolTargetRow <- theTable$event == eventName
                self$eventTable[boolTargetRow, ] <- newRow
            } else {
                #append new row
                self$eventTable <- rbind(theTable,newRow)
            }
            invisible(self)
        },
        #stop timeR
        stop = function(eventName, RMSE = NA_real_, MAE = NA_real_, stars = NA_character_, 
                        params = NA_character_, comment = NA_character_, printmsg = TRUE){
            theTable <- self$eventTable
            verbose  <- self$verbose
            current_time <- format(Sys.time(), "%H:%M:%OS")
            # current_time <- as.character(lubridate::now())
            #detect if event already exists
            if (any(theTable$event %in% eventName)){
                #detect if end time for event already exist
                # end_exist <- !is.na(as.character(
                    # theTable[theTable$event ==eventName,][["end"]]))
                # if (end_exist){
                    # out_msg <- paste0("Event: '",eventName,
                                      # "' already has a record.",
                                      # " Overwriting previous one.\n")
                    # private$slprint(out_msg)
                # }
                #modify the end anyway
                isEventRow <- theTable$event == eventName
                startTime <- self$eventTable[isEventRow, ][["start"]]
                duration <- round(as.numeric(difftime(
                    strptime(current_time, format = "%H:%M:%OS"),
                    strptime(startTime, format = "%H:%M:%OS"),
                    units = "secs"
                    )), 3)
                # duration <- difftime(
                    # lubridate::ymd_hms(current_time,tz = Sys.timezone()),
                    # lubridate::ymd_hms(startTime,tz = Sys.timezone()),
                    # units = "secs"
                    # )
                self$eventTable[isEventRow, ][["end"]] <- current_time
                self$eventTable[isEventRow, ][["duration"]] <- duration
                self$eventTable[isEventRow, ][["RMSE"]] <- RMSE
                self$eventTable[isEventRow, ][["MAE"]]  <- MAE
                self$eventTable[isEventRow, ][["stars"]] <- stars
                self$eventTable[isEventRow, ][["params"]] <- params
                self$eventTable[isEventRow, ][["comment"]] <- comment
            } else {
                stop("Event: '",eventName,"'",
                     " doesn't exist. Record won't be created.\n")
            }
            if (printmsg) {
                out_msg <- paste0("For event: '", eventName,
                                  "', ", duration,
                                  " seconds elapsed.\n")
                private$slprint(out_msg)
            }
            invisible(self)
        },
        getTimer = function(...){
            return(self$eventTable)
        },
        removeEvent = function(eventName){
            theTable <- self$eventTable
            boolRow <- !(theTable$event == eventName)
            if(all(boolRow)) {
                out_msg <- paste0("Event '",eventName,
                                  "' doesn't exist. No record is deleted.\n")
                private$slprint(out_msg)
            } else {
                self$eventTable <- theTable[boolRow,]
            }
            invisible(self)
        },
        toggleVerbose = function(...){
            self$verbose = !self$verbose
            out_msg <- paste0("Verbose set to: ",as.character(self$verbose),
                              ".\n")
            writeLines(out_msg)
            invisible(self)
        },
        getStartTime = function(eventName){
            rowIndex <- which(eventName == self$eventTable$event)
            if (length(rowIndex) == 0) {
                stop("event doesn't exist.")
            }
            result <- self$eventTable[rowIndex,"start"]
            return(result)
        },
        getStopTime = function(eventName){
            rowIndex <- which(eventName == self$eventTable$event)
            if (length(rowIndex) == 0) {
                stop("event doesn't exist.")
            }
            result <- self$eventTable[rowIndex,"stop"]
            return(result)
        },
        getDuration = function(eventName){
            rowIndex <- which(eventName == self$eventTable$event)
            if (length(rowIndex) == 0) {
                stop("event doesn't exist.")
            }
            result <- self$eventTable[rowIndex,"duration"]
            return(result)
        },
        getRMSE = function(eventName){
            rowIndex <- which(eventName == self$eventTable$event)
            if (length(rowIndex) == 0) {
                stop("event doesn't exist.")
            }
            result <- self$eventTable[rowIndex,"RMSE"]
            return(result)
        },
        getMAE = function(eventName){
            rowIndex <- which(eventName == self$eventTable$event)
            if (length(rowIndex) == 0) {
                stop("event doesn't exist.")
            }
            result <- self$eventTable[rowIndex,"MAE"]
            return(result)
        },
        getStars = function(eventName){
            rowIndex <- which(eventName == self$eventTable$event)
            if (length(rowIndex) == 0) {
                stop("event doesn't exist.")
            }
            result <- self$eventTable[rowIndex,"stars"]
            return(result)
        },
        getParams = function(eventName){
            rowIndex <- which(eventName == self$eventTable$event)
            if (length(rowIndex) == 0) {
                stop("event doesn't exist.")
            }
            result <- self$eventTable[rowIndex,"params"]
            return(result)
        },
        getComment = function(eventName){
            rowIndex <- which(eventName == self$eventTable$event)
            if (length(rowIndex) == 0) {
                stop("event doesn't exist.")
            }
            result <- self$eventTable[rowIndex,"comment"]
            return(result)
        },
        getEvent = function(eventName){
            rowIndex <- which(eventName == self$eventTable$event)
            if (length(rowIndex) == 0) {
                stop("event doesn't exist.")
            }
            result <- self$eventTable[rowIndex,]
            return(result)
        },
        print = function(...){
            writeLines("Your Table is:")
            print(self$eventTable)
        }
    ),
    #function used to determine whether message is printed
    private = list(
        slprint = function(msg,flag = self$verbose){
            if(flag) writeLines(msg)
        }
    ),
    #%H:%M original. %H:%M:%S PK.
    active = list(now = function(){
        # as.POSIXct(Sys.time(), format = "%Y-%m-%d %H:%M")
        as.POSIXct(Sys.time(), format = "%Y-%m-%d %H:%M:%S")
    })
)


